dlib/control/approximate_linear_models_abstract.h

```diff
@@ -56,41 +59,29 @@ namespace dlib
                 - returns the dimensionality of the PSI() feature vector.  
         !*/
 
-        action_type find_best_action (
-            const state_type& state,
-            const matrix<double,0,1>& w
```

Comment:
This seems confusing.   I see that the lspi code (in dlib/test/lspi.cpp) still defines feature extractors this way.  So this is wrong isn't it?  This function still needs to be here.

A better way to organize this model/feature extractor stuff is probably to have the documentation say there is just one class, not two.  Then you can define two different versions of the model class.  There is the basic one which was previously defined by the example_feature_extractor.  Then there is the more detailed one with the new things you want to add to models.  The new one has all the functions the feature extractor had, but with some additional new ones.  Then when you use any model class it's not like it depends on some feature extractor class.  It's just one class.

After all, the model is the thing the user makes.  So it's weird that later on it's defined to take a feature extractor as a template argument.

Anyway, my point is that it's really unclear what a model or feature extractor is right now.  Clarifying that is the most important thing here since that's the interface a user of the software is expected to implement.  If it's not really clear people will just not use it and this whole RL module will never be used.  Making that model interface simple so people can easily implement it is the central challenge.